docs/design/2023-09-22-tiflash-query-level-auto-spill.md

```diff
@@ -0,0 +1,55 @@
+
+### When to trigger auto spill
+
+In operator's spill, spill is triggered when opeartor's memory usage exceeds the threshold, it does not work for query level's spill, because the query level's memory usage threshold is a hard limit, that is to say if the memory usage exceeds the threshold, TiFlash will throw error and abort the query immediately. So for query level auto spill, it has to trigger auto spill check before the memory usage is too large. In the implementation, there is a new variable to control when to trigger auto spill check: [auto_memory_revoke_trigger_threshold](https://github.com/pingcap/tiflash/blob/ce24115c147293df10a9d9f1c745eb0e81f0f7b5/dbms/src/Interpreters/Settings.h#L317). If the memory usage exceeds `auto_memory_revoke_trigger_threshold * query_memory_usage_limit`, TiFlash will trigger auto spill check, and try to mark operator to spill.
+
+### How to trigger auto spill check
+
+In operator's spill, auto spill check is triggered after each calculation of a input block in each opeartor that support spill. For query level auto spill, the basic idea is almost the same: check auto spill after each calculation of a input block. However, the difference is for query level auto spill, each operator, even if it does not support spill, will consume memory during calculation, so unlike operator's spill, in order to support query level auto spill, auto spill check should be trigger in all operators. In order to avoid add trigger code inside each operator, the implementation add this trigger code to the runtime framework
+
+* For pull mode, auto spill check will be triggered after each `read` in [IProfilingBlockInputStream](https://github.com/pingcap/tiflash/blob/ce24115c147293df10a9d9f1c745eb0e81f0f7b5/dbms/src/DataStreams/IProfilingBlockInputStream.cpp#L122)
+* For push mode, auto spill check will be triggered after each `read/write/tryOutput/transform` in [Operator](https://github.com/pingcap/tiflash/blob/ce24115c147293df10a9d9f1c745eb0e81f0f7b5/dbms/src/Operators/Operator.h)
+
+### When to spill
+
+There are two possible choices
+
+* Synchronous way: Once auto spill check found out an operator need to spill, it stops current work, and spill synchronously
+* Asynchronous way: Seperate spill check and spill data, auto spill trigger only need to mark the opeartor to spill, and for each operator, it need to check the spill flag and spill data if needed
+
+The synchronous way can release memory as soon as possible, but it requires that spill can be triggered at any time by any other threads, this will need a lot of synchronization mechanism, which will cause performance degradation and will greatly increase the probability of deadlock during runtime, so current implementation choose the async way. 
```

Comment:
```suggestion
The synchronous way can release memory as soon as possible, but it requires that spill can be triggered at any time by any other threads, this will need a lot of synchronization mechanism, which will cause performance degradation and will greatly increase the probability of deadlock during runtime, so current implementation chooses the async way. 
```